src/compiler/mod.rs

```diff
@@ -100,31 +130,86 @@ impl<'a> Compiler<'a> {
         }
     }
 
+    fn push_to_datatable(&mut self, value: PactType<'a>) -> Result<(), CompileErr> {
+        self.data_table.push(value);
+        if self.data_table.len() > MAX_ENTRIES {
+            Err(CompileErr::DataTableFull)
+        } else {
+            Ok(())
+        }
+    }
+
     /// Compile an assertion AST node
     fn compile_assertion(&mut self, assertion: &'a ast::Assertion) -> Result<(), CompileErr> {
-        let comparator_op = compile_comparator(&assertion.2)?;
-        self.bytecode.push(comparator_op.into());
-
-        let lhs_load_op = self.compile_subject(&assertion.0)?;
-        match lhs_load_op {
-            OpCode::LD_INPUT(index) | OpCode::LD_USER(index) => {
-                self.bytecode.push(lhs_load_op.into());
-                self.bytecode.push(index);
-            }
-            _ => panic!("unreachable"),
+        let lhs_load = self.compile_subject(&assertion.0)?;
+        let (comparator_op, invert) = compile_comparator(&assertion.1, &assertion.2)?;
+        let rhs_load = self.compile_subject(&assertion.3)?;
+
+        // Determine the Load Order
+        let (load, flip) = match lhs_load.1 {
```

Comment:
If we allow user and input data to appear in any order, we don't need to do 2 possible flips.
i.e. and here we only need to return error in 1 case:
if lhs_load.1 == DataTable && rhs_load.1 == DataTable) return Err(...)

Reply:
Since we only use a single bit in indicate INPUT_VS_INPUT or INPUT_VS_USER, and we only encode the index to the data table, the Load op has to be in the correct order. Therefore my suggestion would not work.